bundles/org.eclipse.mita.base/src/org/eclipse/mita/base/typesystem/solver/Substitution.xtend

```diff
@@ -60,61 +66,115 @@ class Substitution {
 	}
 	
 	def void add(Map<TypeVariable, AbstractType> content) {
-		if(false && content.entrySet.exists[it.key.toString == "f_13"]) {
-			print("")
-		}
 		val newContent = new Substitution();
-		newContent.content = content;
-		val resultSub = newContent.apply(this);
-		this.content = resultSub.content;
-		//this.checkConsistency();
+		content.forEach[tv, typ|
+			newContent.addToContent(tv, typ);
+		]
+		newContent.applyMutating(this);
 	}
 	def void add(Iterable<Pair<TypeVariable, AbstractType>> content) {
-		
 		add(content.toMap([it.key], [it.value]))
 	}
 	
 	def Substitution replace(TypeVariable from, AbstractType with) {
 		val result = new Substitution();
-		result.content = new HashMap(content.mapValues[it.replace(from, with)])
+		result.constraintSystemProvider = constraintSystemProvider;
+		result.content = new Int2ObjectOpenHashMap(content.mapValues[it.replace(from, with)])
+		// nothing changes for typevariable idx
+		result.idxToTypeVariable = new Int2ObjectOpenHashMap(idxToTypeVariable);
+		return result;
+	}
+	def Substitution replaceMutating(TypeVariable from, AbstractType with) {
+		val result = this;
+		for(int k: result.content.keySet.force) {
+			val vOld = result.content.get(k);
+			val vNew = vOld.replace(from, with);
+			if(vOld !== vNew) {	
+				result.content.put(k, vNew);	
+			}
+		}
+		// nothing changes for typevariable idx
 		return result;
 	}
 	
 	def AbstractType apply(TypeVariable typeVar) {
 		var AbstractType result = typeVar;
-		var nextResult = content.get(result); 
+		var nextResult = content.get(typeVar.idx); 
 		while(nextResult !== null && result != nextResult && !result.freeVars.empty) {
 			result = nextResult;
 			nextResult = applyToType(result);
 		}
 		return result;
 	}
 	
-	def checkConsistency() {
-		val freeTypeVars = new HashSet(content.values.flatMap[it.freeVars.map[toString]].toSet);
-		val typeVars = content.keySet.map[toString].toSet;
-		freeTypeVars.retainAll(typeVars);
-		if(!freeTypeVars.empty) {
-			return false;
-		}
-		return true;
+	def Substitution apply(Substitution oldEntries) {
+		return new Substitution(this).applyMutating(oldEntries);
 	}
 	
-	def Substitution apply(Substitution oldEntries) {
-		val result = new Substitution();
+	// returns the mutated argument (or a copy of this if other is an empty substitution)
+	def Substitution applyMutating(Substitution oldEntries) {
+		if(oldEntries == EMPTY) {
+			return new Substitution(this);
+		}
+		val result = oldEntries;
 		val newEntries = this;
-		result.content = new HashMap(((newEntries.content.size + oldEntries.content.size) * 1.4) as int);
 		result.constraintSystemProvider = newEntries.constraintSystemProvider ?: oldEntries.constraintSystemProvider;
-		result.content.putAll(oldEntries.content.mapValues[it.replace(newEntries)]);
-		
+		result.idxToTypeVariable.putAll(newEntries.idxToTypeVariable);
+
+		if(newEntries.content.size <= 100) {
```

Comment:
Where does this magic number `100` come from? Is this some boundary of fastutil?
Please add a comment explaining it's purpose and reasoning.